Try to support Hyper-V

guyschaos opened the pull request:
@ailispaw  Please help to review. Thanks.

ryohuang committed:
Try to support Hyper-V

ailispaw commented:
@guyschaos Thank you so much! Looks good.

ailispaw commented on configs/kernel.config:
```diff
@@ -1978,9 +1978,9 @@ CONFIG_VIRTIO_MMIO_CMDLINE_DEVICES=y
-CONFIG_HYPERV_BALLOON=m
+CONFIG_HYPERV=y
+CONFIG_HYPERV_UTILS=y
+CONFIG_HYPERV_BALLOON=y
```
I'm wondering whether these 3 config should be `y` to run.

ailispaw replied:
CONFIG_HYPERV seems mandatory.

ailispaw commented:
Just for my reference:
HyperV Kernel Configuration - Funtoo
http://www.funtoo.org/HyperV_Kernel_Configuration

ailispaw commented:
I built an ISO with your commit + the following modifications.

```diff
--- a/configs/kernel.config
+++ b/configs/kernel.config
@@ -1979,8 +1979,8 @@ CONFIG_VIRTIO_MMIO_CMDLINE_DEVICES=y
 # Microsoft Hyper-V guest support
 #
 CONFIG_HYPERV=y
-CONFIG_HYPERV_UTILS=y
-CONFIG_HYPERV_BALLOON=y
+CONFIG_HYPERV_UTILS=m
+CONFIG_HYPERV_BALLOON=m

 #
 # Xen driver support
```

[barge.iso.zip](https://github.com/bargees/barge-os/files/842749/barge.iso.zip)

@guyschaos Could you test this ISO? And please provide the output of `lsmod` and `dmesg` in the VM.
Thanks.

guyschaos commented:
OK. Testing now. I will feedback ASAP.

guyschaos commented:
Hi @ailispaw , it works! 
![barge](https://cloud.githubusercontent.com/assets/246671/23931417/03b736e6-096d-11e7-8bbe-5fe41e26e896.PNG)

ailispaw commented:
@guyschaos Thank you so much!!
Could you give me the output of `lsmod` and `dmesg` in the VM?

ailispaw commented:
How about an additional disk?
Can you mount a virtual HDD?

guyschaos commented:
@ailispaw Sorry... Forgot lsmod and dmesg

```bash
[bargee@barge ~]$ lsmod
Module                  Size  Used by    Not tainted
xt_nat                  1536 13
veth                    3648  0
ipt_MASQUERADE           960 17
nf_nat_masquerade_ipv4     1412  1 ipt_MASQUERADE
xfrm_user              16512  1
iptable_nat             1280  3
nf_conntrack_ipv4       5248  7
nf_defrag_ipv4          1152  1 nf_conntrack_ipv4
nf_nat_ipv4             3266  1 iptable_nat
xt_addrtype             2240  8
xt_conntrack            2496  6
nf_nat                  9484  3 xt_nat,nf_nat_masquerade_ipv4,nf_nat_ipv4
nf_conntrack           48992  5 nf_nat_masquerade_ipv4,nf_conntrack_ipv4,nf_nat_ipv4,xt_conntrack,nf_nat
[bargee@barge ~]$ dmesg
Linux version 4.9.13-barge (root@7c2f70dab242) (gcc version 5.4.0 (Buildroot 2017.02) ) #1 SMP Tue Mar 14 21:13:42 UTC 2017
Command line: quiet barge.switch_root initrd=/boot/initrd BOOT_IMAGE=/boot/bzImage
x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
x86/fpu: Enabled xstate features 0x7, context size is 832 bytes, using 'standard' format.
x86/fpu: Using 'eager' FPU context switches.
e820: BIOS-provided physical RAM map:
BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
BIOS-e820: [mem 0x000000000009fc00-0x000000000009ffff] reserved
BIOS-e820: [mem 0x00000000000e0000-0x00000000000fffff] reserved
BIOS-e820: [mem 0x0000000000100000-0x00000000f7feffff] usable
BIOS-e820: [mem 0x00000000f7ff0000-0x00000000f7ffefff] ACPI data
BIOS-e820: [mem 0x00000000f7fff000-0x00000000f7ffffff] ACPI NVS
BIOS-e820: [mem 0x0000000100000000-0x0000000278ffffff] usable
NX (Execute Disable) protection: active
SMBIOS 2.3 present.
DMI: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS 090006  04/28/2016
Hypervisor detected: Microsoft HyperV
HyperV: features 0x2e7f, hints 0xc2c
HyperV: LAPIC Timer Frequency: 0x1e8480
clocksource: hyperv_clocksource: mask: 0xffffffffffffffff max_cycles: 0x24e6a1710, max_idle_ns: 440795202120 ns
tsc: Marking TSC unstable due to running on Hyper-V
e820: update [mem 0x00000000-0x00000fff] usable ==> reserved
e820: remove [mem 0x000a0000-0x000fffff] usable
e820: last_pfn = 0x279000 max_arch_pfn = 0x400000000
MTRR default type: uncachable
MTRR fixed ranges enabled:
  00000-9FFFF write-back
  A0000-DFFFF uncachable
  E0000-FFFFF write-back
MTRR variable ranges enabled:
  0 base 0000000000 mask 7F00000000 write-back
  1 base 0100000000 mask 7000000000 write-back
  2 disabled
  3 disabled
  4 disabled
  5 disabled
  6 disabled
  7 disabled
x86/PAT: Configuration [0-7]: WB  WC  UC- UC  WB  WC  UC- WT
e820: last_pfn = 0xf7ff0 max_arch_pfn = 0x400000000
found SMP MP-table at [mem 0x000ff780-0x000ff78f] mapped at [ffff8800000ff780]
Base memory trampoline at [ffff880000099000] 99000 size 24576
Using GB pages for direct mapping
BRK [0x019ce000, 0x019cefff] PGTABLE
BRK [0x019cf000, 0x019cffff] PGTABLE
BRK [0x019d0000, 0x019d0fff] PGTABLE
BRK [0x019d1000, 0x019d1fff] PGTABLE
RAMDISK: [mem 0x7f5b4000-0x7fffefff]
ACPI: Early table checksum verification disabled
ACPI: RSDP 0x00000000000F56F0 000014 (v00 ACPIAM)
ACPI: RSDT 0x00000000F7FF0000 000040 (v01 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: FACP 0x00000000F7FF0200 000081 (v02 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: DSDT 0x00000000F7FF1724 0033BE (v01 MSFTVM MSFTVM02 00000002 INTL 02002026)
ACPI: FACS 0x00000000F7FFF000 000040
ACPI: WAET 0x00000000F7FF1480 000028 (v01 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: SLIC 0x00000000F7FF14C0 000176 (v01 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: OEM0 0x00000000F7FF16C0 000064 (v01 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: SRAT 0x00000000F7FF0600 000150 (v02 VRTUAL MICROSFT 00000001 MSFT 00000001)
ACPI: APIC 0x00000000F7FF0300 000252 (v01 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: OEMB 0x00000000F7FFF040 000064 (v01 VRTUAL MICROSFT 04001628 MSFT 00000097)
ACPI: Local APIC address 0xfee00000
Zone ranges:
  DMA      [mem 0x0000000000001000-0x0000000000ffffff]
  DMA32    [mem 0x0000000001000000-0x00000000ffffffff]
  Normal   [mem 0x0000000100000000-0x0000000278ffffff]
Movable zone start for each node
Early memory node ranges
  node   0: [mem 0x0000000000001000-0x000000000009efff]
  node   0: [mem 0x0000000000100000-0x00000000f7feffff]
  node   0: [mem 0x0000000100000000-0x0000000278ffffff]
Initmem setup node 0 [mem 0x0000000000001000-0x0000000278ffffff]
On node 0 totalpages: 2559886
  DMA zone: 64 pages used for memmap
  DMA zone: 21 pages reserved
  DMA zone: 3998 pages, LIFO batch:0
  DMA32 zone: 16320 pages used for memmap
  DMA32 zone: 1011696 pages, LIFO batch:31
  Normal zone: 24128 pages used for memmap
  Normal zone: 1544192 pages, LIFO batch:31
ACPI: PM-Timer IO Port: 0x408
ACPI: Local APIC address 0xfee00000
ACPI: LAPIC_NMI (acpi_id[0xff] dfl dfl lint[0x1])
IOAPIC[0]: apic_id 0, version 17, address 0xfec00000, GSI 0-23
ACPI: INT_SRC_OVR (bus 0 bus_irq 0 global_irq 2 dfl dfl)
ACPI: INT_SRC_OVR (bus 0 bus_irq 9 global_irq 9 high level)
ACPI: IRQ0 used by override.
ACPI: IRQ9 used by override.
Using ACPI (MADT) for SMP configuration information
smpboot: 64 Processors exceeds NR_CPUS limit of 8
smpboot: Allowing 8 CPUs, 0 hotplug CPUs
e820: [mem 0xf8000000-0xffffffff] available for PCI devices
Booting paravirtualized kernel on bare hardware
clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
setup_percpu: NR_CPUS:8 nr_cpumask_bits:8 nr_cpu_ids:8 nr_node_ids:1
percpu: Embedded 34 pages/cpu @ffff880278c00000 s98968 r8192 d32104 u262144
pcpu-alloc: s98968 r8192 d32104 u262144 alloc=1*2097152
pcpu-alloc: [0] 0 1 2 3 4 5 6 7
Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 2519353
Kernel command line: quiet barge.switch_root initrd=/boot/initrd BOOT_IMAGE=/boot/bzImage
PID hash table entries: 4096 (order: 3, 32768 bytes)
Dentry cache hash table entries: 2097152 (order: 12, 16777216 bytes)
Inode-cache hash table entries: 1048576 (order: 11, 8388608 bytes)
Memory: 9965380K/10239544K available (4527K kernel code, 501K rwdata, 1072K rodata, 856K init, 460K bss, 274164K reserved, 0K cma-reserved)
Hierarchical RCU implementation.
        Build-time adjustment of leaf fanout to 64.
NR_IRQS:4352 nr_irqs:488 16
Console: colour VGA+ 80x25
console [tty0] enabled
tsc: PIT calibration matches PMTIMER. 1 loops
tsc: Detected 3392.280 MHz processor
Calibrating delay loop (skipped), value calculated using timer frequency.. 6784.56 BogoMIPS (lpj=33922800)
pid_max: default: 32768 minimum: 301
ACPI: Core revision 20160831
ACPI: 1 ACPI AML tables successfully acquired and loaded
Security Framework initialized
Mount-cache hash table entries: 32768 (order: 6, 262144 bytes)
Mountpoint-cache hash table entries: 32768 (order: 6, 262144 bytes)
CPU: Physical Processor ID: 0
CPU: Processor Core ID: 0
Last level iTLB entries: 4KB 1024, 2MB 1024, 4MB 1024
Last level dTLB entries: 4KB 1024, 2MB 1024, 4MB 1024, 1GB 4
Freeing SMP alternatives memory: 20K (ffffffff81955000 - ffffffff8195a000)
smpboot: Max logical packages: 8
..TIMER: vector=0x30 apic1=0 pin1=2 apic2=-1 pin2=-1
smpboot: CPU0: Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz (family: 0x6, model: 0x3c, stepping: 0x3)
Performance Events: unsupported p6 CPU model 60 no PMU driver, software events only.
x86: Booting SMP configuration:
.... node  #0, CPUs:      #1 #2 #3 #4 #5 #6 #7
x86: Booted up 1 node, 8 CPUs
smpboot: Total of 8 processors activated (54428.15 BogoMIPS)
devtmpfs: initialized
clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
futex hash table entries: 2048 (order: 5, 131072 bytes)
NET: Registered protocol family 16
cpuidle: using governor ladder
cpuidle: using governor menu
ACPI: bus type PCI registered
PCI: Using configuration type 1 for base access
HugeTLB registered 2 MB page size, pre-allocated 0 pages
ACPI: Added _OSI(Module Device)
ACPI: Added _OSI(Processor Device)
ACPI: Added _OSI(3.0 _SCP Extensions)
ACPI: Added _OSI(Processor Aggregator Device)
ACPI: Interpreter enabled
ACPI: (supports S0 S5)
ACPI: Using IOAPIC for interrupt routing
PCI: Using host bridge windows from ACPI; if necessary, use "pci=nocrs" and report a bug
ACPI: PCI Root Bridge [PCI0] (domain 0000 [bus 00-ff])
acpi PNP0A03:00: _OSC: OS supports [ASPM ClockPM Segments MSI]
acpi PNP0A03:00: _OSC failed (AE_NOT_FOUND); disabling ASPM
PCI host bridge to bus 0000:00
pci_bus 0000:00: root bus resource [mem 0xfe0000000-0xfffffffff window]
pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000bffff window]
pci_bus 0000:00: root bus resource [mem 0xf8000000-0xfffbffff window]
pci_bus 0000:00: root bus resource [bus 00-ff]
pci 0000:00:00.0: [8086:7192] type 00 class 0x060000
pci 0000:00:07.0: [8086:7110] type 00 class 0x060100
pci 0000:00:07.1: [8086:7111] type 00 class 0x010180
pci 0000:00:07.1: reg 0x20: [io  0xffa0-0xffaf]
pci 0000:00:07.1: legacy IDE quirk: reg 0x10: [io  0x01f0-0x01f7]
pci 0000:00:07.1: legacy IDE quirk: reg 0x14: [io  0x03f6]
pci 0000:00:07.1: legacy IDE quirk: reg 0x18: [io  0x0170-0x0177]
pci 0000:00:07.1: legacy IDE quirk: reg 0x1c: [io  0x0376]
pci 0000:00:07.3: [8086:7113] type 00 class 0x068000
* Found PM-Timer Bug on the chipset. Due to workarounds for a bug,
* this clock source is slow. Consider trying other clock sources
pci 0000:00:07.3: quirk: [io  0x0400-0x043f] claimed by PIIX4 ACPI
pci 0000:00:08.0: [1414:5353] type 00 class 0x030000
pci 0000:00:08.0: reg 0x10: [mem 0xf8000000-0xfbffffff]
pci_bus 0000:00: on NUMA node 0
ACPI: PCI Interrupt Link [LNKA] (IRQs 3 4 5 7 9 10 *11 12 14 15)
ACPI: PCI Interrupt Link [LNKB] (IRQs 3 4 5 7 9 10 11 12 14 15) *0, disabled.
ACPI: PCI Interrupt Link [LNKC] (IRQs 3 4 5 7 9 10 11 12 14 15) *0, disabled.
ACPI: PCI Interrupt Link [LNKD] (IRQs 3 4 5 7 9 10 11 12 14 15) *0, disabled.
ACPI: Enabled 1 GPEs in block 00 to 0F
vgaarb: setting as boot device: PCI:0000:00:08.0
vgaarb: device added: PCI:0000:00:08.0,decodes=io+mem,owns=io+mem,locks=none
vgaarb: loaded
vgaarb: bridge control possible 0000:00:08.0
SCSI subsystem initialized
libata version 3.00 loaded.
clocksource: hyperv_clocksource_tsc_page: mask: 0xffffffffffffffff max_cycles: 0x24e6a1710, max_idle_ns: 440795202120 ns
hv_vmbus: Hyper-V Host Build:14393-10.0-0-0.576; Vmbus version:4.0
PCI: Using ACPI for IRQ routing
PCI: pci_cache_line_size set to 64 bytes
e820: reserve RAM buffer [mem 0x0009fc00-0x0009ffff]
e820: reserve RAM buffer [mem 0xf7ff0000-0xf7ffffff]
e820: reserve RAM buffer [mem 0x279000000-0x27bffffff]
clocksource: Switched to clocksource hyperv_clocksource_tsc_page
pnp: PnP ACPI init
pnp 00:00: Plug and Play ACPI device, IDs PNP0b00 (active)
pnp 00:01: Plug and Play ACPI device, IDs PNP0303 PNP030b (active)
pnp 00:02: Plug and Play ACPI device, IDs PNP0f03 PNP0f13 (active)
pnp 00:03: [dma 0 disabled]
pnp 00:03: Plug and Play ACPI device, IDs PNP0501 (active)
pnp 00:04: [dma 0 disabled]
pnp 00:04: Plug and Play ACPI device, IDs PNP0501 (active)
pnp 00:05: [dma 2]
pnp 00:05: Plug and Play ACPI device, IDs PNP0700 (active)
system 00:06: [io  0x01e0-0x01ef] has been reserved
system 00:06: [io  0x0160-0x016f] has been reserved
system 00:06: [io  0x0278-0x027f] has been reserved
system 00:06: [io  0x0378-0x037f] has been reserved
system 00:06: [io  0x0678-0x067f] has been reserved
system 00:06: [io  0x0778-0x077f] has been reserved
system 00:06: [io  0x04d0-0x04d1] has been reserved
system 00:06: Plug and Play ACPI device, IDs PNP0c02 (active)
system 00:07: [io  0x0400-0x043f] has been reserved
system 00:07: [io  0x0370-0x0371] has been reserved
system 00:07: [io  0x0440-0x044f] has been reserved
system 00:07: [mem 0xfec00000-0xfec00fff] could not be reserved
system 00:07: [mem 0xfee00000-0xfee00fff] has been reserved
system 00:07: Plug and Play ACPI device, IDs PNP0c02 (active)
system 00:08: [mem 0x00000000-0x0009ffff] could not be reserved
system 00:08: [mem 0x000c0000-0x000dffff] could not be reserved
system 00:08: [mem 0x000e0000-0x000fffff] could not be reserved
system 00:08: [mem 0x00100000-0xf7ffffff] could not be reserved
system 00:08: [mem 0xfffc0000-0xffffffff] has been reserved
system 00:08: Plug and Play ACPI device, IDs PNP0c01 (active)
pnp: PnP ACPI: found 9 devices
clocksource: acpi_pm: mask: 0xffffff max_cycles: 0xffffff, max_idle_ns: 2085701024 ns
pci_bus 0000:00: resource 4 [mem 0xfe0000000-0xfffffffff window]
pci_bus 0000:00: resource 5 [io  0x0000-0x0cf7 window]
pci_bus 0000:00: resource 6 [io  0x0d00-0xffff window]
pci_bus 0000:00: resource 7 [mem 0x000a0000-0x000bffff window]
pci_bus 0000:00: resource 8 [mem 0xf8000000-0xfffbffff window]
NET: Registered protocol family 2
TCP established hash table entries: 131072 (order: 8, 1048576 bytes)
TCP bind hash table entries: 65536 (order: 8, 1048576 bytes)
TCP: Hash tables configured (established 131072 bind 65536)
UDP hash table entries: 8192 (order: 6, 262144 bytes)
UDP-Lite hash table entries: 8192 (order: 6, 262144 bytes)
NET: Registered protocol family 1
RPC: Registered named UNIX socket transport module.
RPC: Registered udp transport module.
RPC: Registered tcp transport module.
RPC: Registered tcp NFSv4.1 backchannel transport module.
pci 0000:00:00.0: Limiting direct PCI/PCI transfers
pci 0000:00:08.0: Video device with shadowed ROM at [mem 0x000c0000-0x000dffff]
PCI: CLS 0 bytes, default 64
Trying to unpack rootfs image as initramfs...
random: fast init done
Freeing initrd memory: 10540K (ffff88007f5b4000 - ffff88007ffff000)
PCI-DMA: Using software bounce buffering for IO (SWIOTLB)
software IO TLB [mem 0xf3ff0000-0xf7ff0000] (64MB) mapped at [ffff8800f3ff0000-ffff8800f7feffff]
RAPL PMU: API unit is 2^-32 Joules, 4 fixed counters, 10737418240 ms ovfl timer
RAPL PMU: hw unit of domain pp0-core 2^-0 Joules
RAPL PMU: hw unit of domain package 2^-0 Joules
RAPL PMU: hw unit of domain dram 2^-0 Joules
RAPL PMU: hw unit of domain pp1-gpu 2^-0 Joules
clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x30e5d1e8143, max_idle_ns: 440795301366 ns
workingset: timestamp_bits=46 max_order=22 bucket_order=0
NFS: Registering the id_resolver key type
Key type id_resolver registered
Key type id_legacy registered
nfs4filelayout_init: NFSv4 File Layout Driver Registering...
fuse init (API version 7.26)
9p: Installing v9fs 9p2000 file system support
Block layer SCSI generic (bsg) driver version 0.4 loaded (major 254)
io scheduler noop registered
io scheduler deadline registered (default)
io scheduler cfq registered
intel_idle: does not run on family 6 model 60
input: Power Button as /devices/LNXSYSTM:00/LNXPWRBN:00/input/input0
ACPI: Power Button [PWRF]
xenfs: not registering filesystem on non-xen platform
Serial: 8250/16550 driver, 4 ports, IRQ sharing disabled
00:03: ttyS0 at I/O 0x3f8 (irq = 4, base_baud = 115200) is a 16550A
00:04: ttyS1 at I/O 0x2f8 (irq = 3, base_baud = 115200) is a 16550A
Hangcheck: starting hangcheck timer 0.9.1 (tick is 180 seconds, margin is 60 seconds).
brd: module loaded
loop: module loaded
megasas: 06.811.02.00-rc1
VMware PVSCSI driver - version 1.0.7.0-k
hv_vmbus: registering driver hv_storvsc
scsi host0: storvsc_host_t
scsi 0:0:0:0: Direct-Access              ST1000DM003-1ER1 CC46 PQ: 0 ANSI: 0
scsi host0: scsi scan: INQUIRY result too short (5), using 36
scsi 0:0:0:1: Direct-Access                                    PQ: 0 ANSI: 0
scsi 0:0:0:2: Direct-Access                                    PQ: 0 ANSI: 0
scsi 0:0:0:3: Direct-Access                                    PQ: 0 ANSI: 0
scsi 0:0:0:4: Direct-Access                                    PQ: 0 ANSI: 0
scsi 0:0:0:5: Direct-Access                                    PQ: 0 ANSI: 0
scsi 0:0:0:6: Direct-Access                                    PQ: 0 ANSI: 0
scsi 0:0:0:7: Direct-Access                                    PQ: 0 ANSI: 0
scsi host1: storvsc_host_t
scsi 1:0:0:0: Direct-Access     Msft     Virtual Disk     1.0  PQ: 0 ANSI: 5
Ethernet Channel Bonding Driver: v3.7.1 (April 27, 2011)
sd 0:0:0:0: [sda] 1953525168 512-byte logical blocks: (1.00 TB/932 GiB)
sd 0:0:0:1: [sdb] Sector size 0 reported, assuming 512.
sd 0:0:0:1: [sdb] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:1: [sdb] 0-byte physical blocks
sd 0:0:0:1: [sdb] Write Protect is off
sd 0:0:0:1: [sdb] Mode Sense: 00 00 00 00
sd 0:0:0:1: [sdb] Asking for cache data failed
sd 0:0:0:4: [sde] Sector size 0 reported, assuming 512.
sd 0:0:0:4: [sde] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:4: [sde] 0-byte physical blocks
tun: Universal TUN/TAP device driver, 1.6
tun: (C) 1999-2004 Max Krasnyansky <maxk@qualcomm.com>
e1000: Intel(R) PRO/1000 Network Driver - version 7.3.21-k8-NAPI
e1000: Copyright (c) 1999-2006 Intel Corporation.
VMware vmxnet3 virtual NIC driver - version 1.4.a.0-k-NAPI
hv_vmbus: registering driver hv_netvsc
sd 0:0:0:0: [sda] Write Protect is off
sd 0:0:0:0: [sda] Mode Sense: 0f 00 00 00
sd 0:0:0:4: [sde] Write Protect is off
sd 0:0:0:4: [sde] Mode Sense: 00 00 00 00
sd 0:0:0:1: [sdb] Assuming drive cache: write through
sd 0:0:0:2: [sdc] Sector size 0 reported, assuming 512.
sd 0:0:0:2: [sdc] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:2: [sdc] 0-byte physical blocks
sd 0:0:0:3: [sdd] Sector size 0 reported, assuming 512.
sd 0:0:0:3: [sdd] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:3: [sdd] 0-byte physical blocks
hv_netvsc: hv_netvsc channel opened successfully
sd 0:0:0:2: [sdc] Write Protect is off
sd 0:0:0:2: [sdc] Mode Sense: 00 00 00 00
sd 0:0:0:2: [sdc] Asking for cache data failed
sd 0:0:0:2: [sdc] Assuming drive cache: write through
sd 0:0:0:3: [sdd] Write Protect is off
sd 0:0:0:3: [sdd] Mode Sense: 00 00 00 00
sd 0:0:0:3: [sdd] Asking for cache data failed
sd 0:0:0:3: [sdd] Assuming drive cache: write through
sd 1:0:0:0: [sdi] 16777216 512-byte logical blocks: (8.59 GB/8.00 GiB)
sd 1:0:0:0: [sdi] 4096-byte physical blocks
sd 0:0:0:3: [sdd] Sector size 0 reported, assuming 512.
sd 0:0:0:5: [sdf] Sector size 0 reported, assuming 512.
sd 0:0:0:5: [sdf] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:5: [sdf] 0-byte physical blocks
sd 0:0:0:4: [sde] Asking for cache data failed
sd 0:0:0:4: [sde] Assuming drive cache: write through
sd 0:0:0:7: [sdh] Sector size 0 reported, assuming 512.
sd 0:0:0:7: [sdh] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:7: [sdh] 0-byte physical blocks
sd 0:0:0:5: [sdf] Write Protect is off
sd 0:0:0:5: [sdf] Mode Sense: 00 00 00 00
sd 0:0:0:6: [sdg] Sector size 0 reported, assuming 512.
sd 0:0:0:6: [sdg] 1 512-byte logical blocks: (512 B/512 B)
sd 0:0:0:6: [sdg] 0-byte physical blocks
sd 0:0:0:2: [sdc] Sector size 0 reported, assuming 512.
sd 0:0:0:6: [sdg] Write Protect is off
sd 0:0:0:6: [sdg] Mode Sense: 00 00 00 00
sd 0:0:0:6: [sdg] Asking for cache data failed
sd 0:0:0:6: [sdg] Assuming drive cache: write through
sd 0:0:0:7: [sdh] Write Protect is off
sd 0:0:0:7: [sdh] Mode Sense: 00 00 00 00
sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
sd 0:0:0:5: [sdf] Asking for cache data failed
sd 0:0:0:5: [sdf] Assuming drive cache: write through
sd 0:0:0:7: [sdh] Asking for cache data failed
sd 0:0:0:7: [sdh] Assuming drive cache: write through
sd 0:0:0:4: [sde] Sector size 0 reported, assuming 512.
sd 0:0:0:6: [sdg] Sector size 0 reported, assuming 512.
sd 0:0:0:3: [sdd] Sector size 0 reported, assuming 512.
sd 0:0:0:5: [sdf] Sector size 0 reported, assuming 512.
sd 0:0:0:7: [sdh] Sector size 0 reported, assuming 512.
sd 0:0:0:2: [sdc] Sector size 0 reported, assuming 512.
sd 0:0:0:3: [sdd] Attached SCSI disk
sd 0:0:0:2: [sdc] Attached SCSI disk
sd 0:0:0:4: [sde] Sector size 0 reported, assuming 512.
sd 0:0:0:6: [sdg] Sector size 0 reported, assuming 512.
sd 0:0:0:6: [sdg] Attached SCSI disk
sd 0:0:0:4: [sde] Attached SCSI disk
sd 0:0:0:5: [sdf] Sector size 0 reported, assuming 512.
sd 0:0:0:7: [sdh] Sector size 0 reported, assuming 512.
sd 0:0:0:5: [sdf] Attached SCSI disk
sd 1:0:0:0: [sdi] Write Protect is off
sd 1:0:0:0: [sdi] Mode Sense: 0f 00 00 00
sd 0:0:0:7: [sdh] Attached SCSI disk
sd 1:0:0:0: [sdi] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
sd 0:0:0:1: [sdb] Sector size 0 reported, assuming 512.
sd 0:0:0:1: [sdb] Attached SCSI disk
 sdi: sdi1
hv_netvsc b019b8d6-a5c7-44e7-b7e1-6d34f30f19ee: Send section size: 6144, Section count:2560
sd 1:0:0:0: [sdi] Attached SCSI disk
hv_netvsc b019b8d6-a5c7-44e7-b7e1-6d34f30f19ee: Device MAC 00:15:5d:00:01:01 link state up
hv_netvsc: hv_netvsc channel opened successfully
hv_netvsc 0e8195a1-c4bc-4283-94b3-bbf4965137cd: Send section size: 6144, Section count:2560
hv_netvsc 0e8195a1-c4bc-4283-94b3-bbf4965137cd: Device MAC 00:15:5d:00:01:02 link state up
Fusion MPT base driver 3.04.20
Copyright (c) 1999-2008 LSI Corporation
Fusion MPT SAS Host driver 3.04.20
i8042: PNP: PS/2 Controller [PNP0303:PS2K,PNP0f03:PS2M] at 0x60,0x64 irq 1,12
serio: i8042 KBD port at 0x60,0x64 irq 1
serio: i8042 AUX port at 0x60,0x64 irq 12
hv_vmbus: registering driver hyperv_keyboard
Netfilter messages via NETLINK v0.30.
nfnl_acct: registering with nfnetlink.
ip_tables: (C) 2000-2006 Netfilter Core Team
NET: Registered protocol family 10
NET: Registered protocol family 17
NET: Registered protocol family 15
Bridge firewalling registered
9pnet: Installing 9P2000 support
Key type dns_resolver registered
microcode: sig=0x306c3, pf=0x1, revision=0xffffffff
microcode: Microcode Update Driver: v2.01 <tigran@aivazian.fsnet.co.uk>, Peter Oruba
registered taskstats version 1
input: AT Translated Set 2 keyboard as /devices/platform/i8042/serio0/input/input1
input: AT Translated Set 2 keyboard as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A03:00/device:07/VMBUS:01/d34b2567-b9b6-42b9-8778-0a4ec0b955bf/serio2/input/input3
 sda: sda1
sd 0:0:0:0: [sda] Attached SCSI disk
Freeing unused kernel memory: 856K (ffffffff8187f000 - ffffffff81955000)
Write protecting the kernel read-only data: 8192k
Freeing unused kernel memory: 1604K (ffff88000146f000 - ffff880001600000)
Freeing unused kernel memory: 976K (ffff88000170c000 - ffff880001800000)
EXT4-fs (sda1): mounted filesystem with ordered data mode. Opts: (null)
Adding 8386556k swap on /dev/sdi1.  Priority:-1 extents:1 across:8386556k FS
random: crng init done
nf_conntrack version 0.5.0 (65536 buckets, 262144 max)
Initializing XFRM netlink socket
IPv6: ADDRCONF(NETDEV_UP): br-fff4ce0d6249: link is not ready
IPv6: ADDRCONF(NETDEV_UP): docker0: link is not ready
IPv6: ADDRCONF(NETDEV_UP): br-411b76499f9a: link is not ready
IPv6: ADDRCONF(NETDEV_UP): br-6ab17f388208: link is not ready
IPv6: ADDRCONF(NETDEV_UP): br-6dbaf2c11c08: link is not ready
IPv6: ADDRCONF(NETDEV_UP): br-869bbf3fc7d2: link is not ready
br-fff4ce0d6249: port 1(vethaf948ab) entered blocking state
br-fff4ce0d6249: port 1(vethaf948ab) entered disabled state
device vethaf948ab entered promiscuous mode
IPv6: ADDRCONF(NETDEV_UP): vethaf948ab: link is not ready
br-fff4ce0d6249: port 1(vethaf948ab) entered blocking state
br-fff4ce0d6249: port 1(vethaf948ab) entered forwarding state
br-fff4ce0d6249: port 1(vethaf948ab) entered disabled state
br-411b76499f9a: port 1(veth6605b96) entered blocking state
br-411b76499f9a: port 1(veth6605b96) entered disabled state
device veth6605b96 entered promiscuous mode
IPv6: ADDRCONF(NETDEV_UP): veth6605b96: link is not ready
br-411b76499f9a: port 1(veth6605b96) entered blocking state
br-411b76499f9a: port 1(veth6605b96) entered forwarding state
br-411b76499f9a: port 1(veth6605b96) entered disabled state
eth0: renamed from vetheec15a9
IPv6: ADDRCONF(NETDEV_CHANGE): vethaf948ab: link becomes ready
br-fff4ce0d6249: port 1(vethaf948ab) entered blocking state
br-fff4ce0d6249: port 1(vethaf948ab) entered forwarding state
IPv6: ADDRCONF(NETDEV_CHANGE): br-fff4ce0d6249: link becomes ready
eth0: renamed from veth46e7314
IPv6: ADDRCONF(NETDEV_CHANGE): veth6605b96: link becomes ready
br-411b76499f9a: port 1(veth6605b96) entered blocking state
br-411b76499f9a: port 1(veth6605b96) entered forwarding state
IPv6: ADDRCONF(NETDEV_CHANGE): br-411b76499f9a: link becomes ready
IPv6: eth0: IPv6 duplicate address fe80::42:acff:fe12:2 detected!
IPv6: eth0: IPv6 duplicate address fe80::42:acff:fe13:2 detected!
```
And, I've mounted 2 virtual HDDs. One for data, another for SWAP.

```bash
[bargee@barge ~]$ mount
tmpfs on / type tmpfs (rw,relatime,size=8981440k)
devtmpfs on /dev type devtmpfs (rw,relatime,size=4982700k,nr_inodes=1245675,mode=755)
devpts on /dev/pts type devpts (rw,relatime,mode=600,ptmxmode=000)
proc on /proc type proc (rw,relatime)
tmpfs on /run type tmpfs (rw,relatime)
sysfs on /sys type sysfs (rw,relatime)
/dev/sda1 on /mnt/data type ext4 (rw,relatime,data=ordered)
overlay on /etc type overlay (rw,relatime,lowerdir=/etc,upperdir=/mnt/data/etc,workdir=/mnt/data/work/etc)
/dev/sda1 on /mnt/sda1 type ext4 (rw,relatime,data=ordered)
cgroup on /sys/fs/cgroup type tmpfs (rw,relatime)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,relatime,cpuset)
cgroup on /sys/fs/cgroup/cpu type cgroup (rw,relatime,cpu)
cgroup on /sys/fs/cgroup/cpuacct type cgroup (rw,relatime,cpuacct)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,relatime,blkio)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,relatime,memory)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,relatime,devices)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,relatime,freezer)
cgroup on /sys/fs/cgroup/net_cls type cgroup (rw,relatime,net_cls)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,relatime,perf_event)
cgroup on /sys/fs/cgroup/net_prio type cgroup (rw,relatime,net_prio)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,relatime,hugetlb)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,relatime,pids)
/dev/sda1 on /mnt/data/var/lib/docker/overlay type ext4 (rw,relatime,data=ordered)
overlay on /mnt/data/var/lib/docker/overlay/108947ab4aa46cd4fba325d1c24cebc3ab72608cfbc421bdf7ea2442c8bdbef4/merged type overlay (rw,relatime,lowerdir=/mnt/data/var/lib/docker/overlay/9610625f285df79e82e2b8973ac4046038b7a8628d4d65d3b9cd83a152928576/root,upperdir=/mnt/data/var/lib/docker/overlay/108947ab4aa46cd4fba325d1c24cebc3ab72608cfbc421bdf7ea2442c8bdbef4/upper,workdir=/mnt/data/var/lib/docker/overlay/108947ab4aa46cd4fba325d1c24cebc3ab72608cfbc421bdf7ea2442c8bdbef4/work)
overlay on /mnt/data/var/lib/docker/overlay/12c8dcc3ab244081b677d82f896b74501dbed534f93e19b11feb2c7ed0450037/merged type overlay (rw,relatime,lowerdir=/mnt/data/var/lib/docker/overlay/17ed85147e1473dc32d74a7f24a8d4ada547872cdf72e77e964f76227e3c6c5c/root,upperdir=/mnt/data/var/lib/docker/overlay/12c8dcc3ab244081b677d82f896b74501dbed534f93e19b11feb2c7ed0450037/upper,workdir=/mnt/data/var/lib/docker/overlay/12c8dcc3ab244081b677d82f896b74501dbed534f93e19b11feb2c7ed0450037/work)
shm on /mnt/data/var/lib/docker/containers/e520d9d9ba7fc729e151bf769492ca4849b1ec63ece6012a4c821245a7822243/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /mnt/data/var/lib/docker/containers/6b936f0b54926903fbb8e05dad902b0f52d3988c9e19505508e2ffe9c9361ca8/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
nsfs on /var/run/docker/netns/5569acc67f25 type nsfs (rw)
nsfs on /var/run/docker/netns/e91446963f38 type nsfs (rw)

```

```bash
[bargee@barge ~]$ free
             total       used       free     shared    buffers     cached
Mem:       9979376     386348    9593028      43036      38756     135928
-/+ buffers/cache:     211664    9767712
Swap:      8386556          0    8386556
```

Everything looks good for me.

guyschaos commented:
@ailispaw A little problem here.

```bash
[bargee@barge ~]$ sudo pkg install tmux
Downloading...FAIL
Building...
Unable to find image 'ailispaw/barge-pkg:2.4.1-dev' locally
docker: Error response from daemon: manifest for ailispaw/barge-pkg:2.4.1-dev not found.
See 'docker run --help'.
[bargee@barge ~]$ docker pull alpine
Using default tag: latest
latest: Pulling from library/alpine
627beaf3eaaf: Already exists
Digest: sha256:58e1a1bb75db1b5a24a462dd5e2915277ea06438c3f105138f97eb53149673c4
Status: Downloaded newer image for alpine:latest
```

Looks like that "'ailispaw/barge-pkg:2.4.1-dev'" only works with you local build.

ailispaw commented:
You can set `VERSION=2.4.0` in `/etc/default/pkg` for that.
It's a secret workaround. :p

I will release the next version with this PR soon.

Thank you so much for your help.

guyschaos commented:
Yeah. My pleasure. 
Thanks for your great work and waiting for your next release.
